pkg/importer/upload_data_source.go

```diff
@@ -0,0 +1,78 @@
+package importer
+
+import (
+	"io"
+	"net/url"
+	"path/filepath"
+
+	"k8s.io/klog"
+	cdiv1 "kubevirt.io/containerized-data-importer/pkg/apis/core/v1alpha1"
+	"kubevirt.io/containerized-data-importer/pkg/util"
+)
+
+// UploadDataSource contains all the information need to upload data into a data volume.
+// Sequence of phases:
+// 1. Info -> Transfer (In Info phase the format readers are configured)
+// 2. Transfer -> Process
+// 3. Process -> Convert
+type UploadDataSource struct {
+	// Data strean
+	stream io.ReadCloser
+	// stack of readers
+	readers *FormatReaders
+	// url to a file in scratch space.
+	url *url.URL
+}
+
+// NewUploadDataSource creates a new instance of an UploadDataSource
+func NewUploadDataSource(stream io.ReadCloser) *UploadDataSource {
+	return &UploadDataSource{
+		stream: stream,
+	}
+}
+
+// Info is called to get initial information about the data.
+func (ud *UploadDataSource) Info() (ProcessingPhase, error) {
+	var err error
+	// Hardcoded to only accept kubevirt content type.
+	ud.readers, err = NewFormatReaders(ud.stream, cdiv1.DataVolumeKubeVirt)
+	if err != nil {
+		klog.Errorf("Error creating readers: %v", err)
+		return Error, err
+	}
+	return Transfer, nil
```

Comment:
Can this function return `TransferToTarget` if we have detected that the user is not uploading a `qcow2` file?  This could potentially speed things up greatly.  Remember that there may be a real human being at the other side of this upload.  And if that human happens to be one of our qa people in they have the slowest storage in the world.

Reply:
Done